### PR TITLE
[erchef] Fix ACL updates when actor names includes '.'

### LINKED ISSUE #1274: Upgrade from Chef 11 to 12 fails in fix_permissions stage with error 400 for every node
In the final stages of an upgrade from Open Source Chef server 11.1.7 to Chef Server Core 12.5.6 the `chef-server-ctl upgrade` process kicks off a knife script called fix_permissions.knife that is supposed to do some RBAC magic, but this results in an error 400 Bad Request for every node we have on the server.

I have checked whether the clients actually exist on the server after migration, and they do.

```
Open source Chef 11 server data successfully uploaded to Chef 12 server
INFO: Using configuration from /tmp/knife-ec-backup-config.rb
INFO: Starting RBAC Permissions Update
INFO: Fixing Permissions for Clients on Node Objects
INFO: HTTP Request Returned 400 Bad Request: The actor(s) 'foo.example.com' do not exist in this organization as clients or users.
ERROR: Fixing permissions for node 'foo.example.com': 400 "Bad Request"
INFO: HTTP Request Returned 400 Bad Request: The actor(s) 'bar.example.com' do not exist in this organization as clients or users.
ERROR: Fixing permissions for node 'bar.example.com': 400 "Bad Request"
INFO: HTTP Request Returned 400 Bad Request: The actor(s) 'baz.example.com' do not exist in this organization as clients or users.
ERROR: Fixing permissions for node 'baz.example.com': 400 "Bad Request"
[...]
```

The exact command that is run at this stage is `/opt/opscode/embedded/bin/knife exec -V -c /tmp/knife-ec-backup-config.rb /opt/opscode/embedded/service/omnibus-ctl/knife/fix_permissions.knife`

I ran it again manually with more verbosity, this was the output:

```
DEBUG: Chef::HTTP calling Chef::HTTP::JSONInput#handle_request
DEBUG: Chef::HTTP calling Chef::HTTP::JSONOutput#handle_request
DEBUG: Chef::HTTP calling Chef::HTTP::CookieManager#handle_request
DEBUG: Chef::HTTP calling Chef::HTTP::Decompressor#handle_request
DEBUG: Chef::HTTP calling Chef::HTTP::Authenticator#handle_request
DEBUG: Signing the request as pivotal
DEBUG: Chef::HTTP calling Chef::HTTP::RemoteRequestID#handle_request
DEBUG: Chef::HTTP calling Chef::HTTP::APIVersions#handle_request
DEBUG: Chef::HTTP calling Chef::HTTP::ValidateContentLength#handle_request
DEBUG: Initiating GET to https://localhost/nodes/foo.example.com/_acl
DEBUG: ---- HTTP Request Header Data: ----
DEBUG: Accept: application/json
DEBUG: Accept-Encoding: gzip;q=1.0,deflate;q=0.6,identity;q=0.3
DEBUG: X-Ops-Server-API-Version: 1
DEBUG: X-OPS-SIGN: algorithm=sha1;version=1.1;
DEBUG: X-OPS-USERID: pivotal
DEBUG: X-OPS-TIMESTAMP: 2017-05-16T12:17:07Z
DEBUG: X-OPS-CONTENT-HASH: dfgdfsg
DEBUG: X-OPS-AUTHORIZATION-1: sdfgsdf/ugy3Bgsx
DEBUG: X-OPS-AUTHORIZATION-2: sdfgdfs
DEBUG: X-OPS-AUTHORIZATION-3: gsdfgsdfg+sdfgdfg+sdfg
DEBUG: X-OPS-AUTHORIZATION-4: sgsfdg+sdgfsdfg
DEBUG: X-OPS-AUTHORIZATION-5: sdfgdf
DEBUG: X-OPS-AUTHORIZATION-6: sdfgdf/sdfggdsfg+dsfgdf==
DEBUG: HOST: localhost:443
DEBUG: X-REMOTE-REQUEST-ID: c7fcbe5b-c349-4ba4-acd2-4200c29cdc6c
DEBUG: ---- End HTTP Request Header Data ----
DEBUG: ---- HTTP Status and Header Data: ----
DEBUG: HTTP 1.1 200 OK
DEBUG: server: openresty/1.11.2.1
DEBUG: date: Tue, 16 May 2017 12:17:07 GMT
DEBUG: content-type: application/json
DEBUG: transfer-encoding: chunked
DEBUG: connection: close
DEBUG: x-ops-server-api-version: {"min_version":"0","max_version":"1","request_version":"1","response_version":"1"}
DEBUG: x-ops-api-info: flavor=cs;version=12.0.0;oc_erchef=12.15.6
DEBUG: content-encoding: gzip
DEBUG: ---- End HTTP Status/Header Data ----
DEBUG: Chef::HTTP calling Chef::HTTP::ValidateContentLength#handle_response
DEBUG: HTTP server did not include a Content-Length header in response, cannot identify truncated downloads.
DEBUG: Chef::HTTP calling Chef::HTTP::APIVersions#handle_response
DEBUG: Chef::HTTP calling Chef::HTTP::RemoteRequestID#handle_response
DEBUG: Chef::HTTP calling Chef::HTTP::Authenticator#handle_response
DEBUG: Chef::HTTP calling Chef::HTTP::Decompressor#handle_response
DEBUG: Decompressing gzip response
DEBUG: Chef::HTTP calling Chef::HTTP::CookieManager#handle_response
DEBUG: Chef::HTTP calling Chef::HTTP::JSONOutput#handle_response
DEBUG: Chef::HTTP calling Chef::HTTP::JSONInput#handle_response
DEBUG: Chef::HTTP calling Chef::HTTP::JSONInput#handle_request
DEBUG: Chef::HTTP calling Chef::HTTP::JSONOutput#handle_request
DEBUG: Chef::HTTP calling Chef::HTTP::CookieManager#handle_request
DEBUG: Chef::HTTP calling Chef::HTTP::Decompressor#handle_request
DEBUG: Chef::HTTP calling Chef::HTTP::Authenticator#handle_request
DEBUG: Signing the request as pivotal
DEBUG: Chef::HTTP calling Chef::HTTP::RemoteRequestID#handle_request
DEBUG: Chef::HTTP calling Chef::HTTP::APIVersions#handle_request
DEBUG: Chef::HTTP calling Chef::HTTP::ValidateContentLength#handle_request
DEBUG: Initiating PUT to https://localhost/nodes/foo.example.com/_acl/create
DEBUG: ---- HTTP Request Header Data: ----
DEBUG: Content-Type: application/json
DEBUG: Accept: application/json
DEBUG: Accept-Encoding: gzip;q=1.0,deflate;q=0.6,identity;q=0.3
DEBUG: X-Ops-Server-API-Version: 1
DEBUG: X-OPS-SIGN: algorithm=sha1;version=1.1;
DEBUG: X-OPS-USERID: pivotal
DEBUG: X-OPS-TIMESTAMP: 2017-05-16T12:17:07Z
DEBUG: X-OPS-CONTENT-HASH: dfgdfsg
DEBUG: X-OPS-AUTHORIZATION-1: dsfgdf+sdfgdfsg+aasdfasdf+asdfsdf/asdfasdf
DEBUG: X-OPS-AUTHORIZATION-2: asdfasdf/asdfsdaf
DEBUG: X-OPS-AUTHORIZATION-3: asdfsd/2+asdfsad+asdfsd/2b
DEBUG: X-OPS-AUTHORIZATION-4: p+asdfsd/asdfsd/sadfsd
DEBUG: X-OPS-AUTHORIZATION-5: asdfsdaf
DEBUG: X-OPS-AUTHORIZATION-6: asdfsda/asdfsda/asdfsd/asdfsda+fasdfsda==
DEBUG: HOST: localhost:443
DEBUG: X-REMOTE-REQUEST-ID: c7fcbe5b-c349-4ba4-acd2-4200c29cdc6c
DEBUG: Content-Length: 107
DEBUG: ---- End HTTP Request Header Data ----
DEBUG: ---- HTTP Status and Header Data: ----
DEBUG: HTTP 1.1 400 Bad Request
DEBUG: server: openresty/1.11.2.1
DEBUG: date: Tue, 16 May 2017 12:17:07 GMT
DEBUG: content-type: application/json
DEBUG: content-length: 117
DEBUG: connection: close
DEBUG: x-ops-server-api-version: {"min_version":"0","max_version":"1","request_version":"1","response_version":"1"}
DEBUG: x-ops-api-info: flavor=cs;version=12.0.0;oc_erchef=12.15.6
DEBUG: ---- End HTTP Status/Header Data ----
DEBUG: ---- HTTP Response Body ----
DEBUG: {"error":["The actor(s) 'foo.example.com' do not exist in this organization as clients or users."]}
DEBUG: ---- End HTTP Response Body -----
DEBUG: ---- HTTP Request Body ----
DEBUG: {"create":{"actors":["pivotal","foo.example.com"],"groups":["admins","clients","users"]}}
DEBUG: ---- End HTTP Request Body ----
DEBUG: Chef::HTTP calling Chef::HTTP::ValidateContentLength#handle_response
DEBUG: Content-Length validated correctly.
DEBUG: Chef::HTTP calling Chef::HTTP::APIVersions#handle_response
DEBUG: Chef::HTTP calling Chef::HTTP::RemoteRequestID#handle_response
DEBUG: Chef::HTTP calling Chef::HTTP::Authenticator#handle_response
DEBUG: Chef::HTTP calling Chef::HTTP::Decompressor#handle_response
DEBUG: Chef::HTTP calling Chef::HTTP::CookieManager#handle_response
DEBUG: Chef::HTTP calling Chef::HTTP::JSONOutput#handle_response
DEBUG: Chef::HTTP calling Chef::HTTP::JSONInput#handle_response
INFO: HTTP Request Returned 400 Bad Request: The actor(s) 'foo.example.com' do not exist in this organization as clients or users.
ERROR: Fixing permissions for node 'foo.example.com': 400 "Bad Request"
```

I have tried setting the ACL using knife-acl but that also borks with a `HTTP Request Returned 400 Bad Request: The client(s) 'foo.example.com' do not exist in this organization.`

The end result is that clients don't have update privileges for the nodes, and all chef runs fail with a 403 Forbidden.

### PR DESCRIPTION
**Pull request by stevendanna**
Unlike users and groups, client names can contain the `.` character.
In fact, this character this likely since the client name is often
based on the node name which is based on the FQDN of the node.

In chef/chef-server#1159, we added the ability to add name-spaced
actors and groups to ACLs.  However, that changed validated the actor
names against the more restrictive user/group regular expression.

This change uses the standard NAME_REGEX for validation which includes
`.`.  It does mean that some invalid user and group names will pass
validation; however, presumably such users and groups will fail to be
found in the database since they can't be created.

Fixes #1274

Signed-off-by: Steven Danna <steve@chef.io>

### CODE REVIEW

File: oc-chef-pedant/spec/api/account/account_acl_spec.rb [hunk 1]
```diff
@@ -1288,6 +1297,20 @@ def rand_id
...
+                    let(:update_body) do
+                      { permission => {
+                          "actors" => [@client_with_dot.name, platform.admin_user.name, "pivotal"],
+                          "groups" => ["admins"]
+                        }
+                      }
+                    end
+
+                    it "returns 200", :acl do
+                      put("#{request_url}/#{permission}", platform.admin_user, :payload => update_body).should look_like({:status => 200})
```
Comment (marcparadise) [inline oc-chef-pedant/spec/api/account/account_acl_spec.rb]: Let's also verify through a `get` that the name isn't munged in any way,

File: src/oc_erchef/apps/chef_objects/src/chef_regex.erl [hunk 2]
```diff
@@ -58,7 +58,8 @@
 %% http://wiki.chef.io/display/chef/Roles#Roles-name.  Judging
 %% from the cookbook names and recipes in the opscode/cookbooks
 %% repository, this regular expression applies to them as well.
--define(NAME_REGEX, "[.[:alnum:]_-]+").
+-define(NAME_CHAR_CLASS, "[.[:alnum:]_-]").
+-define(NAME_REGEX, ?NAME_CHAR_CLASS ++ "+").
```
Comment (marcparadise) [inline src/oc_erchef/apps/chef_objects/src/chef_regex.erl]: It's minor, but I don't know if the precompile will resolve this - the append might be done inline on each call. Same for the others. May not be significant enough to matter.

Comment (stevendanna) [inline src/oc_erchef/apps/chef_objects/src/chef_regex.erl]: Hm, I wonder if we avoid the ++ and just use the feature where you can put two string literals next to each other: `"foo" "bar"` if that happens at compile time.

Comment (stevendanna) [inline src/oc_erchef/apps/chef_objects/src/chef_regex.erl]: Confirmed it does:
``` > cat test.erl
-module(test).
-export([foo/0]).
foo() ->
    "bar" " " "baz".
> erl
Erlang/OTP 18 [erts-7.3.1] [source] [64-bit] [smp:8:8] [async-threads:10] [hipe] [kernel-poll:false]

Eshell V7.3.1  (abort with ^G)
1> compile:file("test.erl", S).
* 1: variable 'S' is unbound
2> compile:file("test.erl", 'S').
{ok,test}
3>
User switch command
 --> q
> head -15 test.S
{module, test}.  %% version = 0

{exports, [{foo,0},{module_info,0},{module_info,1}]}.

{attributes, []}.

{labels, 7}.


{function, foo, 0, 2}.
  {label,1}.
    {line,[{location,"test.erl",3}]}.
    {func_info,{atom,test},{atom,foo},0}.
  {label,2}.
    {move,{literal,"bar baz"},{x,0}}.
```

**Pull request merged**